### PR TITLE
REF: EA quantile logic to EA._quantile

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -1494,6 +1496,48 @@ def _empty(cls, shape: Shape, dtype: ExtensionDtype):
...
         return result
 
+    def _quantile(
+        self: ExtensionArrayT, qs: npt.NDArray[np.float64], interpolation: str
+    ) -> ExtensionArrayT:
+        """
+        ExtensionArray compatibility layer for quantile_with_mask.
+
+        We pretend that an ExtensionArray with shape (N,) is actually (1, N,)
+        for compatibility with non-EA code.
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/base.py:1506]: This docstring would need to be updated to be written for EA implementors (eg for those it doesn't matter this is a compatility layer for quantile_with_mask)

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
I am personally not super fond of moving this code. The `ExtensionArray` class also serves as a base class and example implementation for external EA authors, and for that purpose I am not sure this is useful (it allows to override it, of course (I don't know if we had feature requests for that up to now?), but for that purpose I would also simplify the base implementation). 

(the dual use case as base class for internal usage vs external usage is sometimes a bit conflicting ..)

Is it really annoying to have the code in array_algos/quantile.py ?

File: pandas/core/arrays/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/base.py:1506]: Good point, will update

**Comment by jbrockmendel** [comment]
> (the dual use case as base class for internal usage vs external usage is sometimes a bit conflicting ..)

i sympathize with this

> Is it really annoying to have the code in array_algos/quantile.py ?

A bit, yes.  The idea of array_algos is to not depend on EA implementation details (the docstring in `array_algos.__init__` hints at this).  More importantly, the special-treatment of pandas-internal (i.e. MaskedArray and a bit more subtly Sparse) is a pattern we've been trying to avoid.  I don't see any other way around this, am open to ideas.

**Review by jreback — CHANGES_REQUESTED** [review]
am fine with this, though anyway to combine some of the added code that is in base & masked?

**Comment by jbrockmendel** [comment]
> though anyway to combine some of the added code that is in base & masked?

there might be, but i haven't found it.  trouble is we call the constructors differently in the masked code

**Review by jreback — APPROVED** [review]

**Pull request merged**